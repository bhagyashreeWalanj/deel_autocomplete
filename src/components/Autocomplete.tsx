import React, { useState, useEffect, useRef } from 'react'
import { TiLocation, TiDelete } from 'react-icons/ti'
import { ISuggestion } from '../interface/ISuggestion'

export interface IAutocomplete {
  countries: ISuggestion[]
}

const Autocomplete = ({ countries }: IAutocomplete) => {
  const [searchValue, setSearchValue] = useState<string>('')
  const [newArray, setNewArray] = useState<ISuggestion[]>([])
  const [showSuggestion, setShowSuggestion] = useState(false)

  const autoCompleteRef = useRef<HTMLDivElement>(null)

  let filteredCountries: ISuggestion[] = []
  if (newArray.length !== 0) {
    filteredCountries = newArray
  } else {
    countries.forEach((country) => {
      if (country.option.toLowerCase().includes(searchValue.toLowerCase())) {
        filteredCountries.push({
          option: country.option,
          highlighted: country.highlighted,
        })
      }
    })
  }

  const handleOnChangeText = (e: any) => {
    let searchText: string = e.target.value
    searchText.length <= 0 ? setShowSuggestion(false) : setShowSuggestion(true)

    setSearchValue(searchText)
    let finalArray: ISuggestion[] = []

    countries.forEach((option) => {
      const element: string = option.option
      let matchStartIndex: number = 0
      if (
        (matchStartIndex = element
          .toLowerCase()
          .indexOf(searchText.toLowerCase())) > -1
      ) {
        let matchStart = element.slice(0, matchStartIndex)
        let highlighted = element.substring(
          matchStartIndex,
          matchStartIndex + searchText.length,
        )
        let matchEnd: string = element.substring(
          matchStartIndex + searchText.length,
        )

        if (matchStartIndex >= 0) {
          let matchText = `${matchStart}<strong>${highlighted}</strong>${matchEnd}`

          finalArray.push({
            option: element,
            highlighted: matchText,
          })
        }
      }
    })
    setNewArray(finalArray)
  }

  // handle onClick of list element to country select from the list and set on input
  const handleSuggestion = (suggestion: string) => {
    setSearchValue(suggestion)
    setShowSuggestion(false)
  }

  // To hide the list if mouse click on screen
  const handleOutsideClick = (event: any) => {
    const { current: wrap } = autoCompleteRef
    if (wrap && !wrap.contains(event.target)) {
      setShowSuggestion(false)
    }
  }

  // delete the text from input
  const handleDelete = () => {
    setSearchValue('')
    setShowSuggestion(true)
  }

  useEffect(() => {
    document.addEventListener('mousedown', handleOutsideClick)
    return () => {
      document.removeEventListener('mousedown', handleOutsideClick)
    }
  }, [])

  return (
    <div className="autocomplete" ref={autoCompleteRef}>
      <input
        type="text"
        id="autocomplete_input"
        value={searchValue}
        onChange={(e) => handleOnChangeText(e)}
        placeholder="Search Country..."
        autoComplete="off"
      />

      {searchValue !== '' ? (
        <TiDelete className="closeButton" onClick={handleDelete} />
      ) : (
        ''
      )}

      {showSuggestion ? (
        filteredCountries.length === 0 ? (
          <ul className="suggestions">
            <li>No Data Found</li>
          </ul>
        ) : (
          <ul className="suggestions">
            {filteredCountries.map((country, index) => {
              return (
                <li
                  className="options"
                  id={`#suggest_${index}`}
                  onClick={() => handleSuggestion(country.option)}
                  key={country.option}
                >
                  <TiLocation style={{ marginRight: '0.8rem' }} />
                  <span
                    dangerouslySetInnerHTML={{ __html: country.highlighted }}
                  ></span>
                </li>
              )
            })}
          </ul>
        )
      ) : (
        ''
      )}
    </div>
  )
}

export default Autocomplete
