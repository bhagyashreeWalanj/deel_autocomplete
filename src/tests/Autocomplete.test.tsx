import Autocomplete from '../components/Autocomplete'
import { mount, shallow, render } from 'enzyme'

describe('Testing Autocomplete component', () => {
  const options = [
    { option: 'India', highlighted: 'India' },
    { option: 'Albania', highlighted: 'Albania' },
    { option: 'Bermuda', highlighted: 'Bermuda' },
    { option: 'Portugal', highlighted: 'Portugal' },
    { option: 'Italy', highlighted: 'Italy' },
    { option: 'Spain', highlighted: 'Spain' },
  ]

  const wrapper = shallow(<Autocomplete countries={options} />)
  const inputContainer = wrapper.find('#autocomplete_input')
  const suggestionList = wrapper.find('.options')

  it('should match the snapshot', () => {
    expect(wrapper).toMatchSnapshot()
  })

  it('should exist the suggestion list', () => {
    expect(suggestionList.exists()).toBeFalsy()
  })

  it('should render all the options as provided', () => {
    expect(suggestionList.length).toBe(0)
  })

  it('should find option in list', () => {
    inputContainer.simulate('change', { target: { value: 'Port' } })
    const filtersuggestion = wrapper.find('.options')
    expect(filtersuggestion.length).toBe(1)
    expect(filtersuggestion.html()).toContain('Port')
  })

  it('should clear the selected text from the input', () => {
    inputContainer.simulate('change', { target: { value: 'Spain' } })
    wrapper.find('.closeButton').simulate('click')
    expect(inputContainer.text()).toBe('')
  })

  it('should highlight the matching text on selection', () => {
    inputContainer.simulate('change', { target: { value: 'Port' } })
    const filtersuggestion = wrapper.find('.options')
    expect(filtersuggestion.find('span').html()).toContain(
      '<strong>Port</strong>ugal',
    )
  })
})
