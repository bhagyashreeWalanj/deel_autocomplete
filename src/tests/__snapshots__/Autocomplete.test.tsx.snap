// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Autocomplete component should match the snapshot 1`] = `
<div
  className="autocomplete"
>
  <input
    autoComplete="off"
    id="autocomplete_input"
    onChange={[Function]}
    placeholder="Search Country..."
    type="text"
    value=""
  />
</div>
`;
